### DISCUSSION TITLE
Porting Amarok Database to ampache

### DISCUSSION THREAD

**Discussion by 1001111github** [ampache/ampache#2838] [Q&A]
Hello All,

Amarok has become increasingly unusable over the past few years, however, I have a large database of information stored in the MySQL database underneath it. I am capable of analyzing the 2 schemas, amarok and ampache, and writing an ETL script but I am hoping someone else has done some of the work already.

TIA
Dave

**Comment by lachlan-00** [comment]
i don't have much knowledge of amarok but i can help out.

Lets break it into parts; I'm assuming you're importing plays, ratings, playlists, library?

**Comment by 1001111github** [comment]
Thank you for offering to help out. Forgive me for the delay in getting back to you. I am very familiar with the Amarok side and I have been spending a lot of time thinking this through since the original collection was mostly vinyl ripped to WAV. There is a lot of custom metadata for these albums, liner notes, release notes, etc.

Playlists - I know I can dump these out as text. I assume I can import these into ampache?
Ratings - This to me is the most critical part. Most of my ratings are not encoded in the mp3 or flac. 
Custom metadata - This data is 99% album related, all I really need is the field and table to use in ampache

The process I am looking at is:
- Scripted conversion of WAV to FLAC including embedding of tags FLAC supports
- Use a new install of ampache to catalog the collection including embedded tags
- Scripted conversion of playlists into ampache
- updating ratings that have not been copied over.

So I think my questions are:
Does anyone have any other ideas on the easiest way to accomplish the above?
Which table is rating data held in?
Which table should I look at for playlist conversion?
Where should  I store album metadata?

TIA
Dave

**Reply by lachlan-00** [reply]
ratings are stored in the rating table.
![image](https://user-images.githubusercontent.com/1305249/113074238-0ff76100-920e-11eb-8762-d11faad1c300.png)

Playlists are in the playlist table
![image](https://user-images.githubusercontent.com/1305249/113075589-93b24d00-9210-11eb-8ed3-0fcc6131a3a2.png)

playlist tracks are stored in playlist_data
![image](https://user-images.githubusercontent.com/1305249/113075631-a9277700-9210-11eb-9cd8-cb03b4369709.png)

It would be easiest to script this with text/TSV exports from amarok and use an api script to import/rate/create

Play history
```
time	songtitle	artistname	albumname	songmbid	artistmbid	albummbid	filename
1523402107	the FLA (Peter Rainman Turned On Mix)	!distain	Raise The Level	9de3e36d-0913-3e4e-a6db-935b79a3649e	c60674c3-1cd8-4b2c-ada3-8f99d94c4aad	1e611cf3-cd75-45d4-acd7-b130b10df5df	/mnt/files-music/albums/!distain/(2007) Raise the Level [SYN100]/206 - The FLA (Peter Rainman Turned on mix)
```

Rating
```
rating	songtitle	artistname	albumname	songmbid	artistmbid	albummbid	filename
5	the FLA (Peter Rainman Turned On Mix)	!distain	Raise The Level	9de3e36d-0913-3e4e-a6db-935b79a3649e	c60674c3-1cd8-4b2c-ada3-8f99d94c4aad	1e611cf3-cd75-45d4-acd7-b130b10df5df	/mnt/files-music/albums/!distain/(2007) Raise the Level [SYN100]/206 - The FLA (Peter Rainman Turned on mix)
```

playlist
```
playlistname	playlisttrack	songtitle	artistname	albumname	songmbid	artistmbid	albummbid	filename
5star tracks	1	the FLA (Peter Rainman Turned On Mix)	!distain	Raise The Level	9de3e36d-0913-3e4e-a6db-935b79a3649e	c60674c3-1cd8-4b2c-ada3-8f99d94c4aad	1e611cf3-cd75-45d4-acd7-b130b10df5df	/mnt/files-music/albums/!distain/(2007) Raise the Level [SYN100]/206 - The FLA (Peter Rainman Turned on mix)
```

I wrote some scripts years ago to import my last.fm history into ampache but it can be updated pretty easily if we use a standard data type and just run through in the api

For album metadata could you explain what that kind of data is?

**Reply by lachlan-00** [reply]
for playlist (m3u, etc) imports it should be handled but i don't use them so i can't actually comment on them.

**Reply by 1001111github** [reply]
Excellent, thank you. I have written a POC connecting to both databases and I am now planning the full conversion. Tjhe album metadata is free-form text. I have tried cramming the data into an existing field and also setting up a private table to hold it. At this point in time, they both work.

**Reply by lachlan-00** [reply]
If you want to put docs or any information up we are happy to share with others. You can also edit the wiki

**Reply by 1001111github** [reply]
So this is/has turned into a massive tag cleanup and transcode to flac chore with other assorted data cleaning tasks along for the ride. Since I am re-writing every tag I though I would check compatibility with ampache before doing all the work. Unfortunately the link on Tag Reading/Meta-Information (https://github.com/ampache/ampache/wiki/Tag-Reading) leads to a "Create New Page" object.

While my specific question is about using the tag POPM, any details on what tags are read by ampache would be helpful.

At this point there is only a single SQL script for ratings. It matches filenames across the 2 databases and
loads the ratings into the ampache database. It will be interesting to see the performance once I start doing large batches.

**Reply by wagnered** [reply]
Ampache currently converts POPM into a user rating.

if `enable_custom_metadata` is set to `true`, all metadata that isn't MusicBrainz is imported into the `metadata` and `metadata_field` tables.  In theory, the same is true for flac(Vorbiscomments) but I haven't tested that theory.

The article at [linux.magazine.com](https://www.linux-magazine.com/Issues/2018/207/FLAC-The-premier-digital-audio-codec) explains how to use the cli `flac` command to convert from wav to flac.   Do not convert to ogg/flac: those files can't be written by the getID3 library which prevents writing back edited songs..

**Reply by 1001111github** [reply]
Excellent, Thank you, I have successfully encoded POPM into mp3s. I can set POPM in flacs but I have not found a client that recognizes it. I will report on my testing with ampache. 
 flac, ffmpeg, mid3v2 are all becoming programs about which I know far too much, lol. Currently, I am  transcoding all the Apple AAC and *m4a to flac. At least they already have tags. 

I will soon have to learn MusicBrainz and set it to work on the wav collection.

**Reply by wagnered** [reply]
POPM is id3v2.* only. Check out [the picard document appendix](https://picard-docs.musicbrainz.org/en/appendices/tag_mapping.html) .  MusicBrainz has a tag with the name: "rating".   I think the Vorbiscomment would be `RATING:user@email = value`. 

Ampache (getID3 library)  should be able to read id3 tags properly embedded in wav files.  I only have a wav file with the basic metadata.  I would like to try a wav file with properly embedded id3 metadata  or any metadata over the basic artist, genre, product and title.

**Reply by 1001111github** [reply]
I have successfully imported POPM and VorbisComment rating tags into ampache. Unfortunately they are attached to user_id = -1 so are not available to anyone including the webside admin user. Updating the user_id (or inserting a duplicate set with a new id) achieves the end result desired.

The alternative is load amarok and ampache into the same mariadb instance and grant the ampache user select privs on the amarok database then running the insert as described above.

Either way, the database updates complete within seconds for 10,000 tracks. The cataloging process with extended metadata enabled took about 4 hours. I am tending to the second solution since if I have to update the database anyways why bother updating all the tags.

Thanks for everyone's help, now I have to learn more about ampache :)
